### ISSUE TITLE
Migration from Junit 4 to Junit 5

### ISSUE THREAD

**Issue by sbernard31** [eclipse/leshan#1381]
I was working on rewriting some tests to be able to test more easily the new transport abstraction layer (#1025).

I was reading some stuff on **Junit 4** and I asked my self what is news in [**Junit 5**](https://junit.org/junit5/) ? 

After reading the documentation, It seems that a lot changes ...  
So before to involve time in test refactoring based on Junit4, we should maybe migrate to **Junit5** first ? :thinking: 

Here some article which gives some reason to migrate to **Junit 5** : 
- https://www.parasoft.com/blog/reaping-the-benefits-of-junit-5/
- https://blog.avenuecode.com/its-time-to-migrate-junit5-is-better-than-junit4

**Comment by sbernard31** [comment]
@jvermillard, @JaroslawLegierski, @Warmek if you have any opinions about this, do not hesitate to share ? :pray:

**Comment by sbernard31** [comment]
Different between Junit4 and 5 are explained [here](https://junit.org/junit5/docs/current/user-guide/#migrating-from-junit4-tips).

I started a branch to experiment how this impacts Leshan tests code base.
See : https://github.com/eclipse/leshan/pull/1391

Most painful tasks for us : 

### 1. change in assert methods

Import should be changed, that's OK.
Less funny argument order changes ...
> Assertions and assumptions in JUnit Jupiter accept the failure message as their last argument instead of the first one.

Hopefully our tests code are not so clean and don't use so much failure message :sweat_smile: 

### 2. Testing exceptions

`@Test(expected = IllegalArgumentException.class)` syntax does not exist anymore.
It  should be replaced by : 
```java
 assertThrowsExactly(IllegalArgumentException.class, () -> {
    // code to tests
});
```

We use it a lot but this no brain and repetitive task. 

### 3. Parameterized Tests

`@RunWith(Parameterized.class)` which allow to parameterize test at class level, does not exist anymore.

If should be replace by [`@ParameterizedTest`](https://junit.org/junit5/docs/current/user-guide/#writing-tests-parameterized-tests) (which is at method/test level)
If parameterize tests at class level is needed, the tricks is to use [something like that](https://stackoverflow.com/a/72783483/5088764).

We use it often and this is clearly the less "no brain" task as you need to rewrite a lot about how you pass arguments.

**Comment by sbernard31** [comment]
`leshan-core` that I used for experiment was the big one.
All other except `leshan-integration-tests` was easy to migrate. (done in PR #1391)

I will try `leshan-integration-tests`this afternoon.

**Comment by sbernard31** [comment]
`leshan-integration-tests` is migrated too.
We was using `LockstepEndpoint` from californium tests maven module, so I had to duplicate it to be able to update it with Junit5 API.
(So I also removed all dependency to californium tests modules)

**One remaining question about [Hamcrest](https://hamcrest.org/JavaHamcrest/) ?** 

Junit4 depends on Hamcrest.  
Junit5 doesn't. (but nothing prevent to use it)

So, I added manually the Hamcrest dependency because we are using it a little in `leshan-integrations-tests` for : 
- 1 call like this : `assertThat(acl.getInstances().keySet(), hasItems(0, 1));`
- lots of calls like : `assertThat(response.getCoapResponse(), is(instanceOf(Response.class)));`
- and also some call in `LockstepEndpoint` class. 

I asking myself if we should keep **Hamcrest** and maybe trying to use it more? 
or this is not so useful and better to remove the dependency ?
